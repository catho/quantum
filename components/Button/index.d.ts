import React from 'react';
import IconButton from './IconButton';

export interface ButtonProps {
    center?: boolean;
    disabled?: boolean;
    stroked?: boolean;
    full?: boolean;
    icon?: string;
    size?: 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge';
    skin?: 'neutral' | 'primary' | 'secondary' | 'success' | 'warning' | 'error';
    type?: 'button' | 'reset' | 'submit';
    children?: React.ReactNode[] | React.ReactNode;
    $as?: JSX.Element | string;
    onClick?: React.MouseEventHandler<HTMLButtonElement>;
    theme?: {
        baseFontSize?: number;
        colors?: object;
        spacing?: object;
        breakpoints?: object;
        components?: {
            button: object;
        };
    };
    id?: string;
}

export default class Button extends React.Component<ButtonProps> {
    static Icon: IconButton;
}
