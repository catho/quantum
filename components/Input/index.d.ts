import { InputHTMLAttributes, MouseEventHandler, ChangeEventHandler, FC } from 'react';
import { Validations } from '../Form';

export type MaskFunction = (rawValue: string) => string[];
export type Mask = boolean | RegExp | string | MaskFunction;
export type Validate =
  | typeof Validations.Required
  | typeof Validations.CPF
  | typeof Validations.CEP
  | typeof Validations.Date
  | typeof Validations.MinLength
  | typeof Validations.MaxLength
  | typeof Validations.Email;
export type CustomValidate =
  | ((param?: { value: string }) => string)
  | { validate: Validate; error: string };

export interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  value?: string;
  label?: string;
  helperText?: string;
  descriptionLabel?: string;
  required?: boolean;
  placeholder?: string;
  type?: 'email' | 'text' | 'tel' | 'number' | 'password' | 'search';
  error?: string | false;
  id?: string;
  name?: string;
  validate?: Validate | CustomValidate | Array<Validate | CustomValidate>;
  mask?: Mask | Mask[];
  onClean?: MouseEventHandler<HTMLInputElement>;
  onChange?: ChangeEventHandler<HTMLInputElement>;
  minLength?: number;
  maxLength?: number;
  skin?: 'default' | 'dark';
  theme?: {
    spacing?: {};
    colors?: {};
    baseFontSize?: number;
  };
}

declare const Input: FC<InputProps> & {
  CEP: FC<InputProps>;
  CNPJ: FC<InputProps>;
  CPF: FC<InputProps>;
  Date: FC<InputProps>;
  Password: FC<InputProps>;
  Phone: FC<InputProps>;
};

export default Input;
