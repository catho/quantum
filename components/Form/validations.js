const msg = {
  REQUIRED: 'Campo obrigatório',
  CPF: 'CPF inválido',
  CEP: 'CEP inválido',
  DATE: 'Data inválida',
  EMAIL: 'E-mail inválido',
  MINLENGTH: (min) => `Mínimo de ${min} caracteres`,
  MAXLENGTH: (max) => `Maximo de ${max} caracteres`,
};

function removeNonDigit(str) {
  return str ? str.replace(/\D+/g, '') : '';
}

class Validations {
  static Required({ value }) {
    return value ? '' : msg.REQUIRED;
  }

  static CPF({ value }, cpf = removeNonDigit(value)) {
    const CPF_SIZE = 11;

    if (!cpf || cpf.length !== CPF_SIZE) {
      return msg.CPF;
    }

    for (let i = 0; i < 10; i += 1) {
      if (cpf === String(i).repeat(CPF_SIZE)) {
        return msg.CPF;
      }
    }

    const validPosition = (position) => {
      let sum = 0;
      let rev;

      for (let i = 0; i < position; i += 1) {
        sum += Number(cpf.charAt(i)) * (position + 1 - i);
      }

      rev = CPF_SIZE - (sum % CPF_SIZE);
      rev = rev === 10 || rev === CPF_SIZE ? 0 : rev;

      return rev === Number(cpf.charAt(position));
    };

    return validPosition(9) && validPosition(10) ? '' : msg.CPF;
  }

  static CEP({ value }) {
    const filtered = String(removeNonDigit(value));

    return filtered.length !== 8 ? msg.CEP : '';
  }

  static Date({ value = '' }) {
    const [day, month, year] = value.split('/');
    const date = new Date(year, month - 1, day);
    const isValidDate = date.getMonth() + 1 === Number(month);

    return isValidDate ? '' : msg.DATE;
  }

  static MinLength({ value = '', minLength }) {
    if (!!minLength && String(value).length < Number(minLength)) {
      return msg.MINLENGTH(minLength);
    }
    return '';
  }

  static MaxLength({ value = '', maxLength }) {
    if (!!maxLength && String(value).length > Number(maxLength)) {
      return msg.MAXLENGTH(maxLength);
    }
    return '';
  }

  static Email({ value }) {
    const pattern =
      /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return pattern.test(value) ? '' : msg.EMAIL;
  }

  static EmailBlockAccent({ value }) {
    const pattern =
      /^(([^<>()[\]\\.,;:\s@À-úûüũÿ"]+(\.[^<>()[\]\\.,;:\s@À-úûüũÿ"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return pattern.test(value) ? '' : msg.EMAIL;
  }
}

export default Validations;
