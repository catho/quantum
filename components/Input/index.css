@import "./Input.module.css";
@import "./sub-components/HelperText.module.css";
@import "./sub-components/InputErrorMessage.module.css";
@import "./sub-components/InputLabel.module.css";
@import "./sub-components/TextInput.module.css";