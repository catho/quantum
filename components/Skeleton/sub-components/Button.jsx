import classNames from 'classnames';
import PropTypes from 'prop-types';
import Rect from './Rect';
import styles from './Button.module.css';

const Button = ({ size = 'medium', className, ...rest }) => {
  const buttonClass = classNames(
    styles.button,
    styles[`button-${size}`],
    className,
  );

  return <Rect {...rest} className={buttonClass} />;
};

Button.propTypes = {
  size: PropTypes.oneOf(['xsmall', 'small', 'medium', 'large']),
};

export default Button;
