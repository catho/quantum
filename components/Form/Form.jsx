import React from 'react';
import PropTypes from 'prop-types';
import Input from '../Input';
import InputTypes from '../Input/InputTypes';

function execValidate(validate, props) {
  if (typeof validate === 'function') {
    return validate(props);
  }

  const { validate: fn, error } = validate;
  const msg = fn(props);

  return msg ? error || msg : '';
}

const typeNames = Object.values(InputTypes).map(
  InputType => InputType.displayName,
);

class Form extends React.Component {
  static _isValidElement = element =>
    React.isValidElement(element) &&
    [Input.displayName, ...typeNames].includes(element.type.displayName);

  constructor(props) {
    super(props);

    this.state = {
      values: {},
      errors: {},
      valid: false,
    };

    const { children } = this.props;
    const { values } = this.state;

    this._recursiveMap(children, child => {
      if (!Form._isValidElement(child)) return;

      const {
        props: { name, value },
      } = child;
      if (value) values[name] = value;
    });
  }

  _createClones = children => {
    const { values, errors } = this.state;

    const generatedClones = this._recursiveMap(children, child => {
      if (!Form._isValidElement(child)) {
        return child;
      }
      const { name, error, onChange = () => {} } = child.props;
      return React.cloneElement(child, {
        value: values[name],
        error: errors[name] || error,
        onChange: e => {
          this._handleChange(e);
          onChange(e);
        },
      });
    });
    return generatedClones;
  };

  _recursiveMap = (children, fn) =>
    React.Children.map(children, child => {
      if (!React.isValidElement(child)) {
        return child;
      }

      if (child.props.children) {
        return fn(
          React.cloneElement(child, {
            children: this._recursiveMap(child.props.children, fn),
          }),
        );
      }
      return fn(child);
    });

  _findError = child => {
    const { props } = child;
    const { values } = this.state;
    const { validate = () => {} } = props;

    let invalid;

    const _props = {
      ...props,
      value: values[props.name],
    };

    if (Array.isArray(validate)) {
      for (let i = 0; i < validate.length; i += 1) {
        invalid = execValidate(validate[i], _props);

        // Stop validation when the first occurs
        if (invalid) break;
      }
    } else {
      invalid = execValidate(validate, _props);
    }

    if (invalid) {
      this.setState({ valid: false });
    }

    return invalid;
  };

  _validateError = children => {
    const { errors } = this.state;

    return this._recursiveMap(children, child => {
      const { name } = child.props;

      const _error = this._findError(child);

      const newError = errors;
      newError[name] = _error;

      this.setState({ errors: newError });
    });
  };

  _handleChange = ({ target: { name, value } }) => {
    const { values, errors } = this.state;
    const newValues = {
      ...values,
      [name]: value,
    };

    const newErrors = errors;
    newErrors[name] = '';

    this.setState({ errors: newErrors });
    this.setState({ values: newValues });
  };

  handleSubmit = event => {
    event.preventDefault();

    const { onSubmit, onValidSubmit, children } = this.props;
    const { errors, values } = this.state;

    this._validateError(children);

    const isValid = !Object.values(errors).find(e => e);

    this.setState({ valid: isValid }, () => {
      const { valid: updatedValid } = this.state;
      onSubmit({ valid: updatedValid });

      if (isValid) onValidSubmit(values);
    });
  };

  render() {
    // Removing invalid form props, to avoid warnings
    const _props = { ...this.props };
    const { children } = _props;
    delete _props.onValidSubmit;

    return (
      <form {..._props} onSubmit={this.handleSubmit}>
        {this._createClones(children)}
      </form>
    );
  }
}

Form.defaultProps = {
  onSubmit: () => {},
  onValidSubmit: () => {},
  noValidate: true,
};

Form.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]).isRequired,
  /** A callback triggered every submit trial, with a flag that indicates that it is valid */
  onSubmit: PropTypes.func,
  /** A callback triggered every valid submit, with a object with input values */
  onValidSubmit: PropTypes.func,
  /** Default html attribute, that prevents default browser validations */
  noValidate: PropTypes.bool,
};

export default Form;
