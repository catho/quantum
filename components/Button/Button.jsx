import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { shadow, hexToRgba, theme as defaultTheme } from '../shared';
import { query } from '../Grid/sub-components/shared';

import {
  components,
  spacing,
  colors as defaultColors,
  baseFontSize as defaultBaseFontSize,
} from '../shared/theme';

import Icon from '../Icon/Icon';

const buttonIconBreakpoints = (props, breakpoint) => {
  const q = query(props.theme.breakpoints)[breakpoint];

  return q`
    margin-right: ${props.theme.spacing[breakpoint] / 2}px;
  `;
};

const ButtonIcon = styled(Icon)`
  pointer-events: none;

  ${props =>
    Object.keys(props.theme.breakpoints).map(breakpoint =>
      buttonIconBreakpoints(props, breakpoint),
    )}
`;

const buttonFontAndLineProps = ({ size, theme: { baseFontSize } }) => {
  const sizes = {
    xsmall: `${baseFontSize * 0.75}px`,
    small: `${baseFontSize * 0.75}px`,
    medium: `${baseFontSize}px`,
    large: `${baseFontSize * 1.25}px`,
    xlarge: `${baseFontSize * 1.5}px`,
  };
  const lineHeights = {
    xsmall: `${baseFontSize * 0.75 * 1.5}px`,
    small: `${baseFontSize * 0.75 * 1.5}px`,
    medium: `${baseFontSize * 1.5}px`,
    large: `${baseFontSize * 1.25 * 1.5}px`,
    xlarge: `${baseFontSize * 1.5 * 1.5}px`,
  };
  return `
    font-size: ${sizes[size]};
    line-height: ${lineHeights[size]};
  `;
};

const StyledButton = styled.button`
  align-items: center;
  display: flex;
  font-weight: bold;
  justify-content: center;
  letter-spacing: 0.2px;
  border-radius: 4px;

  ${props => `cursor: ${props.disabled ? 'not-allowed' : 'pointer'};`}

  ${buttonFontAndLineProps}

  ${({
    size,
    theme: {
      spacing: { large, xlarge, xxlarge, xxxlarge },
    },
  }) => {
    const heights = {
      xsmall: `${large}px`,
      small: `${xlarge}px`,
      medium: `${xxlarge}px`,
      large: `${xxxlarge}px`,
    };

    return `min-height: ${heights[size]};`;
  }}

  ${({
    size,
    theme: {
      spacing: { small, medium, xsmall, xxsmall, large },
    },
  }) => {
    const paddings = {
      xsmall: `${xxsmall}px ${small}px`,
      small: `${small / 2}px ${small}px`,
      medium: `${xsmall}px ${medium}px`,
      large: `${xsmall * 1.25}px ${medium}px`,
      xlarge: `${large / 2}px ${medium}px`,
    };

    return `padding: ${paddings[size]};`;
  }}

  ${props =>
    props.center &&
    `
    margin-left: auto;
    margin-right: auto;
  `}

  transition: all 0.2s ease-in-out;

  ${props => props.full && `width: 100%;`}

  ${({ skin, theme, disabled, stroked }) => {
    const {
      components: {
        button: {
          skins: {
            [skin]: {
              mainColor: {
                100: mainColor100,
                500: mainColor500,
                700: mainColor700,
              },
              text: { 100: text100 },
            },
          },
        },
      },
      colors: {
        neutral: { 100: neutral100, 500: neutral500 },
      },
    } = theme;

    let bgColor;
    let textColor;

    if (disabled && stroked) {
      bgColor = neutral100;
      textColor = neutral500;
    } else if (stroked) {
      bgColor = neutral100;
      textColor = mainColor500;
    } else if (disabled) {
      bgColor = neutral500;
      textColor = neutral100;
    } else {
      bgColor = mainColor500;
      textColor = text100;
    }

    return `
      background-color: ${bgColor};
      color: ${textColor};

      border: 2px solid ${disabled ? neutral500 : mainColor500};

      ${shadow(2, neutral500)({ theme })}

      :hover {
        ${
          !disabled
            ? `
              ${shadow(4, mainColor700)({ theme })}
              background-color: ${stroked ? mainColor100 : mainColor700};
              border-color: ${mainColor700};
            `
            : ''
        }
      }

      :focus,
      :focus-within {
        ${!disabled ? shadow(4, mainColor500)({ theme }) : ''}
      }

      :active {
        ${
          !disabled
            ? `
              ${shadow(8, mainColor700)({ theme })}
              background-color: ${stroked ? mainColor100 : mainColor700};
            `
            : ''
        }
      }

    ${ButtonIcon} {
      ${({
        size,
        theme: {
          spacing: { xxsmall, medium, large },
        },
      }) => {
        const sizes = {
          xsmall: `${medium}px`,
          small: `${medium}px`,
          medium: `${large}px`,
          large: `${large}px`,
          xlarge: `${large}px`,
        };

        return `
          font-size: ${sizes[size]};
          margin-right: ${xxsmall}px;
        `;
      }}
    }
    `;
  }}
`;

const Button = ({ children, icon, size, $as, ...rest }) => (
  <StyledButton as={$as} {...rest} size={size}>
    {icon && <ButtonIcon size={size} name={icon} {...rest} />}
    {children}
  </StyledButton>
);

Button.defaultProps = {
  center: false,
  disabled: false,
  stroked: false,
  full: false,
  icon: '',
  size: 'medium',
  skin: 'primary',
  type: 'button',
  children: undefined,
  $as: undefined,
  onClick: () => {},
  theme: {
    colors: defaultColors,
    baseFontSize: defaultBaseFontSize,
    spacing,
    breakpoints: defaultTheme.breakpoints,
    components: {
      button: components.button,
    },
  },
};

Button.propTypes = {
  center: PropTypes.bool,
  disabled: PropTypes.bool,
  stroked: PropTypes.bool,
  full: PropTypes.bool,
  /** Icon name. The full catalogue can be found
   * [here](/?selectedKind=1.%20Foundation&selectedStory=Icons) */
  icon: PropTypes.string,
  size: PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge']),
  skin: PropTypes.oneOf([
    'neutral',
    'primary',
    'secondary',
    'success',
    'warning',
    'error',
  ]),
  type: PropTypes.oneOf(['button', 'reset', 'submit']),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  /** https://www.styled-components.com/docs/api#as-polymorphic-prop */
  $as: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  onClick: PropTypes.func,
  theme: PropTypes.shape({
    baseFontSize: PropTypes.number,
    colors: PropTypes.object,
    spacing: PropTypes.object,
    breakpoints: PropTypes.arrayOf(PropTypes.any),
    components: PropTypes.shape({
      button: PropTypes.object,
    }),
  }),
};

const IconButton = styled(Button)`
  ${({ skin, theme }) => {
    const {
      components: {
        button: {
          skins: {
            [skin]: {
              mainColor: {
                300: mainColor300,
                500: mainColor500,
                700: mainColor700,
              },
            },
          },
        },
      },
    } = theme;

    return `
    border-radius: 50%;
    border: none;
    color: ${hexToRgba(mainColor500, 0.5)};
    width: 40px;

    background-color: transparent;
    box-shadow: none;
    outline: none;

    ${ButtonIcon} {
      margin-right: 0;
    }

    :hover,
    :focus {
      box-shadow: none;
      background-color: ${hexToRgba(mainColor300, 0.4)};
      color: ${mainColor700};
    }

    :active {
      box-shadow: none;
      background-color: ${hexToRgba(mainColor300, 0.5)};
      color: ${mainColor700};
    }
    `;
  }}
`;

IconButton.propTypes = {
  size: PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge']),
  skin: PropTypes.oneOf([
    'neutral',
    'primary',
    'secondary',
    'success',
    'warning',
    'error',
  ]),
  theme: PropTypes.shape({
    baseFontSize: PropTypes.number,
    colors: PropTypes.object,
    spacing: PropTypes.object,
    components: PropTypes.shape({
      button: PropTypes.object,
    }),
  }),
};

IconButton.defaultProps = {
  size: 'medium',
  skin: 'neutral',
  theme: {
    breakpoints: defaultTheme.breakpoints,
    gutter: '8px',
    baseFontSize: defaultBaseFontSize,
    spacing,
    components: {
      button: components.button,
    },
  },
};

Button.Icon = IconButton;

export default Button;
