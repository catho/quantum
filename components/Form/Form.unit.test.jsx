/* eslint-disable react/display-name */
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import validations from './validations';
import Form from './Form';
import Input from '../Input';
import Button from '../Button';

jest.mock('react-text-mask', () => props => (
  <input type="text" {...{ ...props }} />
));

const onValidSubmitCallback = jest.fn();
const onSubmitCallback = jest.fn();
const mockEvent = { preventDefault: jest.fn() };

const FormWithoutValidations = () => (
  <Form onValidSubmit={onValidSubmitCallback} onSubmit={onSubmitCallback}>
    <Input name="name" label="Name" />
  </Form>
);

const FormWithValidations = () => (
  <Form onValidSubmit={onValidSubmitCallback} onSubmit={onSubmitCallback}>
    <Input name="required" label="required" validate={validations.Required} />
    <Input name="cpf" label="CPF" validate={validations.CPF} />
    <Input name="cep" label="CEP" validate={validations.CEP} />
    <Input name="birthday" label="Birthday" validate={validations.Date} />
    <Input name="email" label="E-mail" validate={validations.Email} />
    <Input
      name="address"
      label="Address"
      validate={validations.MinLength}
      minLength="8"
    />
    <Input
      name="country"
      label="Country"
      validate={validations.MaxLength}
      maxLength="3"
    />

    <Button type="submit"> Enviar </Button>
  </Form>
);

beforeEach(() => {
  onValidSubmitCallback.mockClear();
  onSubmitCallback.mockClear();
});

describe('Form component ', () => {
  it('should match the snapshot', () => {
    expect(renderer.create(FormWithValidations()).toJSON()).toMatchSnapshot();
  });

  describe('Without validations', () => {
    it('Should call "onValidSubmit" and "onSubmit" callback on a valid submit', () => {
      const wrapper = shallow(FormWithoutValidations());
      wrapper.simulate('submit', mockEvent);

      expect(onValidSubmitCallback).toHaveBeenCalled();
      expect(onSubmitCallback).toHaveBeenCalled();
    });
  });

  describe('With validations', () => {
    it('Shouldn\'t call "onValidSubmit" and "onSubmit" callback on a invalid submit', () => {
      const wrapper = shallow(FormWithValidations());
      wrapper.simulate('submit', mockEvent);

      expect(onValidSubmitCallback).not.toHaveBeenCalled();
      expect(onSubmitCallback).toHaveBeenCalled();
    });

    it('Should validate fields', () => {
      const execTest = ({ invalid, valid, validationName, errorMsg }) => {
        const wrapper = shallow(FormWithValidations());
        const getField = () =>
          wrapper.find({ validate: validations[validationName] });

        if (invalid) {
          const initalField = getField();
          initalField.simulate('change', {
            target: { name: initalField.prop('name'), value: invalid },
          });
        }

        wrapper.simulate('submit', mockEvent);

        const beforeChange = getField();
        expect(beforeChange.prop('error')).toBe(errorMsg);

        beforeChange.simulate('change', {
          target: { name: beforeChange.prop('name'), value: valid },
        });

        const afterChange = getField();

        expect(afterChange.prop('error')).toBe('');
      };

      const validationTests = [
        {
          validationName: 'Required',
          errorMsg: 'Campo obrigatório',
          valid: 'foo',
        },
        {
          validationName: 'CPF',
          errorMsg: 'CPF inválido',
          valid: '321.970.213-97',
          invalid: '123.456.212-34',
        },
        {
          validationName: 'CEP',
          errorMsg: 'CEP inválido',
          valid: '02354128',
        },
        {
          validationName: 'Date',
          errorMsg: 'Data inválida',
          valid: '20/11/1981',
          invalid: '30/02/1950',
        },
        {
          validationName: 'Email',
          errorMsg: 'E-mail inválido',
          valid: 'foo@baz.com',
          invalid: 'foo@baz',
        },
        {
          validationName: 'MaxLength',
          errorMsg: 'Maximo de 3 caracteres',
          valid: '123',
          invalid: '123456789',
        },
        {
          validationName: 'MinLength',
          errorMsg: 'Mínimo de 8 caracteres',
          valid: '12345678',
          invalid: '12345',
        },
      ];

      validationTests.forEach(execTest);
    });

    it('Should exec validations in diferent formats', () => {
      const form = (
        <Form onValidSubmit={onValidSubmitCallback} onSubmit={onSubmitCallback}>
          <Input
            name="name"
            label="Name"
            minLength="8"
            validate={[
              validations.Required,
              {
                validate: validations.MinLength,
                error: 'Valor mínimo de caracteres não alcançado',
              },
            ]}
          />
        </Form>
      );

      const wrapper = shallow(form);
      wrapper.simulate('submit', mockEvent);

      expect(onSubmitCallback).toHaveBeenCalled();

      const input = wrapper.find(Input);

      input.simulate('change', {
        target: { name: input.prop('name'), value: 'Some value' },
      });

      wrapper.simulate('submit', mockEvent);

      expect(onValidSubmitCallback).toHaveBeenCalled();
    });
  });

  describe('Valid state', () => {
    it('Should be false when an error occurred', () => {
      const form = (
        <Form onValidSubmit={onValidSubmitCallback} onSubmit={onSubmitCallback}>
          <Input name="foo" validate={validations.Required} />
        </Form>
      );

      const wrapper = shallow(form);
      expect(wrapper.state('valid')).toBe(false);

      wrapper.simulate('submit', mockEvent);

      expect(onSubmitCallback).toHaveBeenCalled();
      expect(onValidSubmitCallback).not.toHaveBeenCalled();
      expect(wrapper.state('valid')).toBe(false);
    });

    it('Should be true when everything is ok', () => {
      const form = (
        <Form onValidSubmit={onValidSubmitCallback} onSubmit={onSubmitCallback}>
          <Input name="foo" validate={validations.Required} value="foo" />
        </Form>
      );

      const wrapper = shallow(form);
      expect(wrapper.state('valid')).toBe(false);

      wrapper.simulate('submit', mockEvent);

      expect(onSubmitCallback).toHaveBeenCalledWith({ valid: true });
      expect(onValidSubmitCallback).toHaveBeenCalledWith({ foo: 'foo' });
      expect(wrapper.state('valid')).toBe(true);
    });
  });
});
