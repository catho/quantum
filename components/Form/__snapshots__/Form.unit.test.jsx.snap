// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component  should match the snapshot 1`] = `
.c0 {
  position: relative;
  margin-bottom: 20px;
  min-width: 250px;
  width: 100%;
}

.c1 {
  display: block;
  font-weight: bold;
  margin: 8px 0 0;
}

.c2 {
  position: relative;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-weight: bold;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 4px;
  cursor: pointer;
  font-size: 16px;
  line-height: 24px;
  min-height: 40px;
  padding: 4px 16px;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  background-color: #1250C4;
  color: #ffffff;
  border: 2px solid #1250C4;
  box-shadow: 0px 3px 1px -2px rgba(153,153,153,0.2),0px 2px 2px 0px rgba(153,153,153,0.14),0px 1px 5px 0px rgba(153,153,153,0.12);
}

.c3:hover {
  box-shadow: 0px 2px 4px -1px rgba(0,47,123,0.2),0px 4px 5px 0px rgba(0,47,123,0.14),0px 1px 10px 0px rgba(0,47,123,0.12);
  background-color: #002F7B;
  border-color: #002F7B;
}

.c3:focus,
.c3:focus-within {
  box-shadow: 0px 2px 4px -1px rgba(18,80,196,0.2),0px 4px 5px 0px rgba(18,80,196,0.14),0px 1px 10px 0px rgba(18,80,196,0.12);
}

.c3:active {
  box-shadow: 0px 5px 5px -3px rgba(0,47,123,0.2),0px 8px 10px 1px rgba(0,47,123,0.14),0px 3px 14px 2px rgba(0,47,123,0.12);
  background-color: #002F7B;
}

.c3 .Button__ButtonIcon-sc-1ovnfsw-0 function (_ref6) {
  -webkit-var: size = _ref6.size, _ref6$theme$spacing = _ref6.theme.spacing, xxsmall = _ref6$theme$spacing.xxsmall, medium = _ref6$theme$spacing.medium, large = _ref6$theme$spacing.large;
  -moz-var: size = _ref6.size, _ref6$theme$spacing = _ref6.theme.spacing, xxsmall = _ref6$theme$spacing.xxsmall, medium = _ref6$theme$spacing.medium, large = _ref6$theme$spacing.large;
  -ms-var: size = _ref6.size, _ref6$theme$spacing = _ref6.theme.spacing, xxsmall = _ref6$theme$spacing.xxsmall, medium = _ref6$theme$spacing.medium, large = _ref6$theme$spacing.large;
  var: size = _ref6.size, _ref6$theme$spacing = _ref6.theme.spacing, xxsmall = _ref6$theme$spacing.xxsmall, medium = _ref6$theme$spacing.medium, large = _ref6$theme$spacing.large;
  return: "\\n          font-size: ".concat(fontSizes[size],";\\n          margin-right: ").concat(xxsmall,"px;\\n        ");
}

.c3 .Button__ButtonIcon-sc-1ovnfsw-0 function (_ref6) var fontSizes = {
  xsmall: "".concat(medium,"px"), small:"".concat(medium,"px"), medium:"".concat(large,"px"), large:"".concat(large,"px"), xlarge:"".concat(large,"px");
}

<form
  noValidate={true}
  onSubmit={[Function]}
>
  <div
    className="c0"
  >
    <label
      className="c1"
      htmlFor="input-0"
    >
      required
    </label>
    
    <div
      className="c2"
    >
      <input
        disabled={false}
        id="input-0"
        mask={[Function]}
        maxLength={255}
        minLength={0}
        name="required"
        onChange={[Function]}
        placeholder=""
        render={[Function]}
        required={false}
        type="text"
        validate={[Function]}
        value=""
      />
      
      
    </div>
    
    
  </div>
  <div
    className="c0"
  >
    <label
      className="c1"
      htmlFor="input-1"
    >
      CPF
    </label>
    
    <div
      className="c2"
    >
      <input
        disabled={false}
        id="input-1"
        mask={[Function]}
        maxLength={255}
        minLength={0}
        name="cpf"
        onChange={[Function]}
        placeholder=""
        render={[Function]}
        required={false}
        type="text"
        validate={[Function]}
        value=""
      />
      
      
    </div>
    
    
  </div>
  <div
    className="c0"
  >
    <label
      className="c1"
      htmlFor="input-2"
    >
      CEP
    </label>
    
    <div
      className="c2"
    >
      <input
        disabled={false}
        id="input-2"
        mask={[Function]}
        maxLength={255}
        minLength={0}
        name="cep"
        onChange={[Function]}
        placeholder=""
        render={[Function]}
        required={false}
        type="text"
        validate={[Function]}
        value=""
      />
      
      
    </div>
    
    
  </div>
  <div
    className="c0"
  >
    <label
      className="c1"
      htmlFor="input-3"
    >
      Birthday
    </label>
    
    <div
      className="c2"
    >
      <input
        disabled={false}
        id="input-3"
        mask={[Function]}
        maxLength={255}
        minLength={0}
        name="birthday"
        onChange={[Function]}
        placeholder=""
        render={[Function]}
        required={false}
        type="text"
        validate={[Function]}
        value=""
      />
      
      
    </div>
    
    
  </div>
  <div
    className="c0"
  >
    <label
      className="c1"
      htmlFor="input-4"
    >
      E-mail
    </label>
    
    <div
      className="c2"
    >
      <input
        disabled={false}
        id="input-4"
        mask={[Function]}
        maxLength={255}
        minLength={0}
        name="email"
        onChange={[Function]}
        placeholder=""
        render={[Function]}
        required={false}
        type="text"
        validate={[Function]}
        value=""
      />
      
      
    </div>
    
    
  </div>
  <div
    className="c0"
  >
    <label
      className="c1"
      htmlFor="input-5"
    >
      Address
    </label>
    
    <div
      className="c2"
    >
      <input
        disabled={false}
        id="input-5"
        mask={[Function]}
        maxLength={255}
        minLength="8"
        name="address"
        onChange={[Function]}
        placeholder=""
        render={[Function]}
        required={false}
        type="text"
        validate={[Function]}
        value=""
      />
      
      
    </div>
    
    
  </div>
  <div
    className="c0"
  >
    <label
      className="c1"
      htmlFor="input-6"
    >
      Country
    </label>
    
    <div
      className="c2"
    >
      <input
        disabled={false}
        id="input-6"
        mask={[Function]}
        maxLength="3"
        minLength={0}
        name="country"
        onChange={[Function]}
        placeholder=""
        render={[Function]}
        required={false}
        type="text"
        validate={[Function]}
        value=""
      />
      
      
    </div>
    
    
  </div>
  <button
    className="c3"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="submit"
  >
    
     Enviar 
  </button>
</form>
`;
